File: MIDTERM-LAB-3/MP10_DetectDuplicates.js
/**
 * MP10 – Detect Duplicate Records
 * Author: Gonzales, Chris Allen S.
 * Course: BSCSIT 1203L 9302-AY225 (Programming 2 - Lab)


 *
 * Description:
 *   Reads the Pearson VUE exam results CSV file and detects duplicate records.
 *   A duplicate is defined as two or more rows that share the same
 *   Candidate name, Exam title, and Exam Date combination.
 *
 * Program Logic:
 *   1. Ask the user for the CSV file path via readline
 *   2. Read the file using Node.js fs module
 *   3. Skip metadata rows until the header row ("Candidate") is found
 *   4. Parse each data row using a quote-aware CSV parser
 *   5. Build a composite key: Candidate + "|" + Exam + "|" + Date
 *   6. Use a Map to group rows by their composite key
 *   7. Report any group that has more than one row (= duplicate)
 *
 * Run with: node MP10_DetectDuplicates.js
 */

// ---- Node.js built-in modules ----
const fs       = require('fs');       // File system: for reading the CSV file
const readline = require('readline'); // For prompting user input in the terminal

// Create a readline interface connected to terminal input/output
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout
});

// Prompt the user to enter the CSV file path
rl.question('Enter the CSV dataset file path: ', (filePath) => {
    rl.close();  // Close the prompt after receiving the input

    // ---------------------------------------------------------------
    // Read the file
    // ---------------------------------------------------------------
    let fileContent;

    try {
        // Read entire file as a UTF-8 string
        fileContent = fs.readFileSync(filePath.trim(), 'utf8');
    } catch (err) {
        console.log('Error: File not found or could not be read – ' + filePath);
        return;
    }

    // Split the file into lines; handle both Windows (\r\n) and Unix (\n) line endings
    const lines = fileContent.split(/\r?\n/);

    // ---------------------------------------------------------------
    // Data containers
    // ---------------------------------------------------------------

    // Map where:
    //   key   = composite string "Candidate|Exam|Date"
    //   value = array of rows (each row is an array of field strings)
    const keyMap = new Map();

    let totalRecords = 0;  // Total valid rows processed
    let headerFound  = false;  // Flag: true once we've passed the header row

    // ---------------------------------------------------------------
    // Parse each line
    // ---------------------------------------------------------------
    for (const line of lines) {

        // Skip metadata lines until the real header row is found
        // The header row is identified by containing the word "Candidate"
        if (!headerFound) {
            if (line.includes('Candidate')) {
                headerFound = true;
            }
            continue;  // Skip this line (metadata or the header itself)
        }

        // Skip blank lines
        if (line.trim() === '') continue;

        // Parse the CSV line into an array of string fields
        const fields = parseCSVLine(line);

        // Skip rows with fewer than 8 columns (not a valid data row)
        if (fields.length < 8) continue;

        // Column index reference:
        // 0 = Candidate, 1 = Type,     2 = Column1(blank),
        // 3 = Exam,      4 = Language, 5 = Date,
        // 6 = Score,     7 = Result,   8 = Time Used
        const candidate = fields[0].trim();
        const exam      = fields[3].trim();
        const date      = fields[5].trim();

        // Skip empty candidate rows (footer lines)
        if (candidate === '') continue;

        totalRecords++;

        // Build the composite key for duplicate detection
        // Format: "CandidateName|ExamTitle|ExamDate"
        const compositeKey = `${candidate}|${exam}|${date}`;

        // Group rows by composite key using the Map
        if (!keyMap.has(compositeKey)) {
            keyMap.set(compositeKey, []);  // Initialize empty array for new key
        }
        keyMap.get(compositeKey).push(fields);  // Add this row to the group
    }

    // Guard: no valid records found
    if (totalRecords === 0) {
        console.log('No valid records found in the file.');
        return;
    }

    // ---------------------------------------------------------------
    // Identify duplicates: groups with more than one row
    // ---------------------------------------------------------------
    const duplicates = [];  // Array of [compositeKey, rowsArray] pairs

    for (const [key, rows] of keyMap.entries()) {
        if (rows.length > 1) {
            duplicates.push([key, rows]);  // This group has duplicates
        }
    }

    const duplicateGroups = duplicates.length;

    // Count total rows that are part of duplicate groups
    let totalDuplicateRows = 0;
    for (const [, rows] of duplicates) {
        totalDuplicateRows += rows.length;
    }

    // ---------------------------------------------------------------
    // Output – Formatted Duplicate Detection Report
    // ---------------------------------------------------------------
    console.log();
    console.log('╔══════════════════════════════════════════════════════════╗');
    console.log('║          MP10 – DUPLICATE RECORDS DETECTION              ║');
    console.log('╚══════════════════════════════════════════════════════════╝');
    console.log('  Source    : Pearson VUE – UPHSD Molino Exam Results');
    console.log('  Duplicate Key: Candidate + Exam + Date');
    console.log('──────────────────────────────────────────────────────────');
    console.log(`  Total Records Scanned       : ${totalRecords}`);
    console.log(`  Duplicate Groups Found      : ${duplicateGroups}`);
    console.log(`  Total Rows in Duplicates    : ${totalDuplicateRows}`);
    console.log('──────────────────────────────────────────────────────────');

    if (duplicateGroups === 0) {
        // No duplicates were found
        console.log('  ✔ No duplicate records were detected in the dataset.');
    } else {
        console.log('  Duplicate Records Found:');
        console.log();

        let groupNumber = 1;  // Label counter for each duplicate group

        for (const [key, rows] of duplicates) {
            // Split the composite key back into its individual parts
            const keyParts       = key.split('|');
            const candidateName  = keyParts[0] || 'N/A';
            const examTitle      = keyParts[1] || 'N/A';
            const examDate       = keyParts[2] || 'N/A';

            console.log(`  [Duplicate Group #${groupNumber++}]`);
            console.log(`    Candidate : ${candidateName}`);
            console.log(`    Exam      : ${examTitle}`);
            console.log(`    Date      : ${examDate}`);
            console.log(`    Count     : ${rows.length} occurrence(s)`);
            console.log('    Rows:');

            // Print each duplicate row's score, result, and time for comparison
            rows.forEach((row, index) => {
                const score  = row[6] ? row[6].trim() : 'N/A';
                const result = row[7] ? row[7].trim() : 'N/A';
                const time   = row[8] ? row[8].trim() : 'N/A';
                console.log(`      Row ${index + 1}: Score=${score.padEnd(6)}  Result=${result.padEnd(6)}  Time=${time}`);
            });

            console.log();
        }
    }

    console.log('══════════════════════════════════════════════════════════');
    console.log('            End of Duplicate Detection Report             ');
    console.log('══════════════════════════════════════════════════════════');
});

// ---------------------------------------------------------------
// Helper function
// ---------------------------------------------------------------

/**
 * parseCSVLine – Parses a single CSV line into an array of string fields.
 *
 * Correctly handles candidate names enclosed in double-quotes like
 * "LastName,FirstName", treating the comma inside quotes as part of
 * the field rather than as a field separator.
 *
 * @param {string} line  - A single raw line from the CSV file
 * @returns {string[]}   - Array of string field values
 */
function parseCSVLine(line) {
    const fields = [];   // Stores each parsed field
    let current  = '';   // Builds up the current field character by character
    let inQuotes = false; // Tracks whether we are inside a double-quoted field

    for (const char of line) {
        if (char === '"') {
            inQuotes = !inQuotes;  // Toggle in/out of quoted mode
        } else if (char === ',' && !inQuotes) {
            fields.push(current);  // Comma outside quotes = field separator
            current = '';          // Reset for the next field
        } else {
            current += char;       // Regular character: add to current field
        }
    }

    fields.push(current);  // Add the final field (no trailing comma)
    return fields;
}

File: MIDTERM-LAB-3/MP10_DetectDuplicates.java
import java.io.*;
import java.util.*;

/**
 * MP10 – Detect Duplicate Records
 * Author: Gonzales, Chris Allen S.
 * Course: BSCSIT 1203L 9302-AY225 (Programming 2 - Lab)


 *
 * Description:
 *   Reads the Pearson VUE exam results CSV file and detects duplicate records.
 *   A duplicate is defined as two or more rows that share the same Candidate name,
 *   Exam title, and Exam Date. This helps identify if any candidate appears to have
 *   taken the same exam more than once on the same day.
 *
 * Program Logic:
 *   1. Ask the user for the CSV file path
 *   2. Skip metadata rows until the header row is found ("Candidate")
 *   3. Parse each data row using a quote-aware CSV parser
 *   4. Build a composite key from: Candidate + Exam + Date
 *   5. Store each key in a map; count how many times each key appears
 *   6. Display all keys that appear more than once (duplicates)
 */
public class MP10_DetectDuplicates {

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        // Prompt the user to enter the path of the CSV file
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        // ---------------------------------------------------------------
        // Data containers
        // ---------------------------------------------------------------

        // Map: composite key -> list of full row data (for display purposes)
        // Key format: "CandidateName|ExamTitle|ExamDate"
        Map<String, List<String[]>> keyMap = new LinkedHashMap<>();

        // Total count of records processed
        int totalRecords = 0;

        // ---------------------------------------------------------------
        // File Reading
        // ---------------------------------------------------------------
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {

            String line;
            boolean headerFound = false;  // Flag: true after the header row is passed

            while ((line = br.readLine()) != null) {

                // Skip metadata lines until the header row containing "Candidate" is found
                if (!headerFound) {
                    if (line.contains("Candidate")) {
                        headerFound = true;
                    }
                    continue;
                }

                // Skip blank lines
                if (line.trim().isEmpty()) continue;

                // Parse the CSV line into an array of fields
                String[] fields = parseCSVLine(line);

                // Ensure the row has at least 8 columns
                if (fields.length < 8) continue;

                // Extract fields by their column index
                String candidate = fields[0].trim();  // Column 0: Candidate name
                // fields[1] = Type (Student/Faculty/NTE)
                // fields[2] = Column1 (blank)
                String exam      = fields[3].trim();  // Column 3: Exam title
                // fields[4] = Language
                String date      = fields[5].trim();  // Column 5: Exam date
                String score     = fields[6].trim();  // Column 6: Score
                String result    = fields[7].trim();  // Column 7: PASS or FAIL

                // Skip empty candidate rows (footer rows)
                if (candidate.isEmpty()) continue;

                totalRecords++;

                // Build a composite key to identify unique exam attempts
                // Format: "CandidateName|ExamTitle|ExamDate"
                String compositeKey = candidate + "|" + exam + "|" + date;

                // Add this row to the map under its composite key
                // If the key doesn't exist yet, create a new list for it
                keyMap.computeIfAbsent(compositeKey, k -> new ArrayList<>())
                      .add(fields);
            }

        } catch (FileNotFoundException e) {
            System.out.println("Error: File not found – " + filePath);
            return;
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }

        // Guard: no records loaded
        if (totalRecords == 0) {
            System.out.println("No valid records found in the file.");
            return;
        }

        // ---------------------------------------------------------------
        // Identify duplicates: keys that appear more than once
        // ---------------------------------------------------------------
        List<Map.Entry<String, List<String[]>>> duplicates = new ArrayList<>();

        for (Map.Entry<String, List<String[]>> entry : keyMap.entrySet()) {
            if (entry.getValue().size() > 1) {
                duplicates.add(entry);  // This key has 2 or more rows = duplicate
            }
        }

        int duplicateGroups = duplicates.size();  // Number of groups with duplicates

        // Count total duplicate rows (all extra appearances beyond the first)
        int totalDuplicateRows = 0;
        for (Map.Entry<String, List<String[]>> entry : duplicates) {
            totalDuplicateRows += entry.getValue().size();  // Count all rows in the group
        }

        // ---------------------------------------------------------------
        // Output – Duplicate Detection Report
        // ---------------------------------------------------------------
        System.out.println();
        System.out.println("╔══════════════════════════════════════════════════════════╗");
        System.out.println("║          MP10 – DUPLICATE RECORDS DETECTION              ║");
        System.out.println("╚══════════════════════════════════════════════════════════╝");
        System.out.println("  Source  : Pearson VUE – UPHSD Molino Exam Results");
        System.out.printf ("  Duplicate Key Used: Candidate + Exam + Date%n");
        System.out.println("──────────────────────────────────────────────────────────");
        System.out.printf ("  Total Records Scanned       : %d%n", totalRecords);
        System.out.printf ("  Duplicate Groups Found      : %d%n", duplicateGroups);
        System.out.printf ("  Total Rows in Duplicates    : %d%n", totalDuplicateRows);
        System.out.println("──────────────────────────────────────────────────────────");

        if (duplicateGroups == 0) {
            // No duplicates found
            System.out.println("  ✔ No duplicate records were detected in the dataset.");
        } else {
            // Display each duplicate group
            System.out.println("  Duplicate Records Found:");
            System.out.println();

            int groupNumber = 1;  // Counter for labeling each duplicate group

            for (Map.Entry<String, List<String[]>> entry : duplicates) {

                // Split the composite key back into its parts for display
                String[] keyParts = entry.getKey().split("\\|");
                String candidateName = keyParts.length > 0 ? keyParts[0] : "N/A";
                String examTitle     = keyParts.length > 1 ? keyParts[1] : "N/A";
                String examDate      = keyParts.length > 2 ? keyParts[2] : "N/A";

                System.out.printf("  [Duplicate Group #%d]%n", groupNumber++);
                System.out.printf("    Candidate : %s%n", candidateName);
                System.out.printf("    Exam      : %s%n", examTitle);
                System.out.printf("    Date      : %s%n", examDate);
                System.out.printf("    Count     : %d occurrence(s)%n", entry.getValue().size());

                // Print each duplicate row's score and result for comparison
                System.out.println("    Rows:");
                int rowNum = 1;
                for (String[] row : entry.getValue()) {
                    String score  = row.length > 6 ? row[6].trim() : "N/A";
                    String result = row.length > 7 ? row[7].trim() : "N/A";
                    String time   = row.length > 8 ? row[8].trim() : "N/A";
                    System.out.printf("      Row %d: Score=%-6s  Result=%-6s  Time=%s%n",
                        rowNum++, score, result, time);
                }
                System.out.println();
            }
        }

        System.out.println("══════════════════════════════════════════════════════════");
        System.out.println("            End of Duplicate Detection Report             ");
        System.out.println("══════════════════════════════════════════════════════════");
    }

    /**
     * parseCSVLine – Parses a single CSV line into an array of string fields.
     *
     * Handles fields enclosed in double-quotes, such as "LastName,FirstName",
     * which is used for candidate names in this dataset. Commas inside
     * quoted fields are treated as part of the field, not as separators.
     *
     * @param line  A raw line of text from the CSV file
     * @return      Array of parsed string fields
     */
    static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder current = new StringBuilder();  // Current field being built
        boolean inQuotes = false;  // True when inside a double-quoted field

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes;  // Toggle quote mode
            } else if (c == ',' && !inQuotes) {
                fields.add(current.toString());  // Save field, reset buffer
                current.setLength(0);
            } else {
                current.append(c);  // Add char to current field
            }
        }
        fields.add(current.toString());  // Save last field

        return fields.toArray(new String[0]);
    }
}
